### PR TITLE
Fix couple memory leaks. Issue #2772.

### PR DESCRIPTION
**Pull request by gtors**

### CODE REVIEW

File: libr/anal/var.c [hunk 1]
```diff
@@ -339,9 +339,7 @@ R_API int r_anal_fcn_var_del_bydelta (RAnal *a, ut64 fna, const char kind, int s
 R_API int r_anal_var_count(RAnal *a, RAnalFunction *fcn, int kind) {
 	char *varlist;
 	int count = 0;
-	RList *list = r_list_new ();
```
Comment (gtors) [inline libr/anal/var.c]: Variable not used and leaking.

File: libr/core/cmd.c [hunk 2]
```diff
@@ -1836,6 +1836,7 @@ R_API int r_core_cmd_foreach3(RCore *core, const char *cmd, char *each) {
 				r_core_cmd0 (core, cmd);
 			}
 			r_core_cmdf (core, "dp %d", origpid);
+			r_list_free (list);
```
Comment (gtors) [inline libr/core/cmd.c]: `dbg->h->threads(...)` - each time creates new list

File: libr/util/buf.c [hunk 3]
```diff
@@ -49,6 +49,9 @@ static RBufferSparse *sparse_append(RList *l, ut64 addr, const ut8 *data, int le
 	s->odata = NULL;
 	s->data = calloc (1, len);
 	if (!data) {
+		if (s->data) {
```
Comment (radare) [inline libr/util/buf.c]: no need to check for null when calling free

File: libr/io/p/io_default.c [hunk 4]
```diff
@@ -103,6 +103,9 @@ RIOMMapFileObj *r_io_def_mmap_create_new_file(RIO  *io, const char *filename, in
 	else mmo->fd = r_sandbox_open (filename, O_RDONLY, mode);
 
 	if (mmo->fd == -1) {
+		if (mmo->filename) {
```
Comment (radare) [inline libr/io/p/io_default.c]: the same. dont check for null pointers when calling free

File: libr/util/buf.c (hunk 3 above)
Comment (ret2libc) [inline libr/util/buf.c]: I think the fix here should be different. We can check for `data` as first thing in the `sparse_append` function and return NULL if !data. After `s->data = calloc (1, len);` there should be a check for `!s->data`, in my opinion.

Comment (radare) [inline libr/util/buf.c]: agree

**Comment by gtors** [comment]
Done.

File: libr/core/asm.c [hunk 5]
```diff
@@ -387,7 +387,13 @@ R_API RList *r_core_asm_bwdisassemble (RCore *core, ut64 addr, int n, int len) {
 	int numinstr, asmlen, ii;
 	RAsmCode *c;
 
-	if (len<1) return NULL;
+	if (len < 1) {
+		if (hits) {
```
Comment (radare) [inline libr/core/asm.c]: there's no problem in r_list_free(NULL) either, so its a double check here

**Comment by gtors** [comment]
Done.

File: libr/util/buf.c [hunk 6]
```diff
@@ -48,7 +49,7 @@ static RBufferSparse *sparse_append(RList *l, ut64 addr, const ut8 *data, int le
 	s->size = len;
 	s->odata = NULL;
 	s->data = calloc (1, len);
-	if (!data) {
+	if (!s->data) {
 		free (s);
 		return NULL;
 	} else {
```
Comment (ret2libc) [inline libr/util/buf.c]: no need for the `else` here since there is the return NULL in the if branch and it makes the code more readable.

**Comment by ret2libc** [comment]
Awesome! Thanks!

**Comment by gtors** [comment]
Done.

**Comment by ret2libc** [comment]
Thanks a lot! Looks good to me :)

**Pull request merged**